File: fsd/perception/occupancy.py
"""Bird's-eye occupancy grid (world frame, log-odds, numpy only).

Cells accumulate log-odds evidence: raycast cells get free evidence,
endpoints get occupied evidence. Positive log-odds = occupied. The grid is
anchored at a world ``origin`` that can be recentered on the ego as it moves.
"""
from __future__ import annotations

import math
from typing import Iterable, Optional, Sequence, Tuple

import numpy as np

from fsd.core.logger import get

log = get("perception.occupancy")


class OccupancyGrid:
    """2-D BEV occupancy grid; cell (0,0) sits at world (ox, oy)."""

    def __init__(self, width_m: float = 120.0, height_m: float = 120.0,
                 resolution: float = 0.25, origin: Tuple[float, float] = (0.0, 0.0)):
        self.res = float(resolution)
        self.w = int(round(width_m / self.res))
        self.h = int(round(height_m / self.res))
        self.ox, self.oy = float(origin[0]), float(origin[1])
        self.lo_hit, self.lo_free = 0.9, -0.4
        self.lo_min, self.lo_max = -4.0, 4.0
        self.grid = np.zeros((self.h, self.w), dtype=np.float32)

    # --------------------------------------------------------------- basics
    def clear(self) -> None:
        self.grid.fill(0.0)

    def set_origin(self, ox: float, oy: float, keep: bool = True) -> None:
        """Move the world anchor (e.g. recenter on ego); optionally keep data."""
        if keep and (abs(ox - self.ox) < self.w * self.res and
                     abs(oy - self.oy) < self.h * self.res):
            dx = int(round((ox - self.ox) / self.res))
            dy = int(round((oy - self.oy) / self.res))
            new = np.zeros_like(self.grid)
            sx0, sx1 = max(0, -dx), min(self.w, self.w - dx)
            sy0, sy1 = max(0, -dy), min(self.h, self.h - dy)
            dx0, dx1 = max(0, dx), min(self.w, self.w + dx)
            dy0, dy1 = max(0, dy), min(self.h, self.h + dy)
            if sx1 > sx0 and sy1 > sy0:
                new[sy0:sy1, sx0:sx1] = self.grid[dy0:dy1, dx0:dx1]
            self.grid = new
        else:
            self.clear()
        self.ox, self.oy = float(ox), float(oy)

    def _cell(self, x: float, y: float) -> Tuple[int, int]:
        return int((x - self.ox) / self.res), int((y - self.oy) / self.res)

    def in_bounds(self, x: float, y: float) -> bool:
        cx, cy = self._cell(x, y)
        return 0 <= cx < self.w and 0 <= cy < self.h

    # --------------------------------------------------------------- update
    def _bump(self, cx: int, cy: int, delta: float) -> None:
        if 0 <= cx < self.w and 0 <= cy < self.h:
            v = self.grid[cy, cx] + delta
            self.grid[cy, cx] = min(max(v, self.lo_min), self.lo_max)

    def mark(self, x: float, y: float, occupied: bool = True) -> None:
        """Add occupied/free evidence for the cell containing world (x, y)."""
        cx, cy = self._cell(x, y)
        self._bump(cx, cy, self.lo_hit if occupied else self.lo_free)

    def mark_points(self, points: Iterable[Sequence[float]],
                    occupied: bool = True) -> None:
        for p in points:
            self.mark(float(p[0]), float(p[1]), occupied)

    def raycast(self, ox: float, oy: float, x: float, y: float,
                hit: bool = True) -> None:
        """Bresenham from (ox,oy) to (x,y); free evidence along the ray,
        occupied evidence at the endpoint if ``hit``."""
        cx0, cy0 = self._cell(ox, oy)
        cx1, cy1 = self._cell(x, y)
        dx, dy = abs(cx1 - cx0), -abs(cy1 - cy0)
        sx = 1 if cx0 < cx1 else -1
        sy = 1 if cy0 < cy1 else -1
        err, cx, cy = dx + dy, cx0, cy0
        while True:
            if cx == cx1 and cy == cy1:
                self._bump(cx, cy, self.lo_hit if hit else self.lo_free)
                break
            self._bump(cx, cy, self.lo_free)
            e2 = 2 * err
            if e2 >= dy:
                err += dy
                cx += sx
            if e2 <= dx:
                err += dx
                cy += sy

    def insert_scan(self, points: Iterable[Sequence[float]],
                    origin: Tuple[float, float],
                    max_range: Optional[float] = None) -> None:
        """Raycast each world-frame scan endpoint from the sensor origin."""
        ox, oy = origin
        for p in points:
            x, y = float(p[0]), float(p[1])
            if max_range is not None and \
                    math.hypot(x - ox, y - oy) > max_range:
                continue
            self.raycast(ox, oy, x, y, hit=True)

    # ---------------------------------------------------------------- query
    def log_odds(self, x: float, y: float) -> float:
        cx, cy = self._cell(x, y)
        if 0 <= cx < self.w and 0 <= cy < self.h:
            return float(self.grid[cy, cx])
        return 0.0

    def is_occupied(self, x: float, y: float, threshold: float = 0.6) -> bool:
        cx, cy = self._cell(x, y)
        return (0 <= cx < self.w and 0 <= cy < self.h and
                self.grid[cy, cx] > threshold)

    def occupied_fraction(self) -> float:
        return float(np.mean(self.grid > 0.6))

    def free_space_ahead(self, ego_x: float, ego_y: float, yaw: float,
                         max_dist: float = 80.0,
                         half_width: float = 1.0) -> float:
        """Meters of clear corridor along ``yaw`` before an occupied cell.

        Marches a corridor of ``2*half_width`` lateral clearance; the first
        occupied cell across the corridor wins. Leaving the grid bounds is
        reported as the distance to the grid edge (unknown beyond).
        """
        ux, uy = math.cos(yaw), math.sin(yaw)
        vx, vy = -uy, ux                          # lateral (right) unit vector
        n_lat = max(1, int(math.ceil(half_width / self.res)))
        lats = np.linspace(-half_width, half_width, 2 * n_lat + 1)
        step = self.res
        s = step
        while s < max_dist:
            bx, by = ego_x + ux * s, ego_y + uy * s
            if not self.in_bounds(bx, by):
                return s                          # ran out of known space
            for lat in lats:
                if self.is_occupied(bx + vx * lat, by + vy * lat):
                    return s
            s += step
        return max_dist

    def as_uint8(self) -> np.ndarray:
        """0 free / 128 unknown / 255 occupied — handy for debugging dumps."""
        return np.clip(128 + self.grid * 64, 0, 255).astype(np.uint8)

File: cpp/src/occupancy.cpp
// Implementation of the BEV occupancy grid — a faithful C++17 port of
// fsd/perception/occupancy.py (numpy arrays -> row-major std::vector<float>).
#include "fsd/occupancy.hpp"

#include <algorithm>
#include <cmath>

namespace fsd {

OccupancyGrid::OccupancyGrid(double width_m, double height_m, double resolution,
                             std::pair<double, double> origin)
    : res(resolution), ox(origin.first), oy(origin.second) {
    w = static_cast<int>(round_nearest(width_m / res));
    h = static_cast<int>(round_nearest(height_m / res));
    grid_.assign(static_cast<std::size_t>(w) * static_cast<std::size_t>(h), 0.0f);
}

// ------------------------------------------------------------------ basics

void OccupancyGrid::clear() { std::fill(grid_.begin(), grid_.end(), 0.0f); }

void OccupancyGrid::set_origin(double new_ox, double new_oy, bool keep) {
    if (keep && std::abs(new_ox - ox) < w * res &&
        std::abs(new_oy - oy) < h * res) {
        const int dx = static_cast<int>(round_nearest((new_ox - ox) / res));
        const int dy = static_cast<int>(round_nearest((new_oy - oy) / res));
        std::vector<float> next(grid_.size(), 0.0f);
        // Python: new[sy0:sy1, sx0:sx1] = grid[dy0:dy1, dx0:dx1]
        const int sx0 = std::max(0, -dx), sx1 = std::min(w, w - dx);
        const int sy0 = std::max(0, -dy), sy1 = std::min(h, h - dy);
        const int dx0 = std::max(0, dx), dx1 = std::min(w, w + dx);
        const int dy0 = std::max(0, dy), dy1 = std::min(h, h + dy);
        for (int syy = sy0, dyy = dy0; syy < sy1 && dyy < dy1; ++syy, ++dyy) {
            for (int sxx = sx0, dxx = dx0; sxx < sx1 && dxx < dx1; ++sxx, ++dxx) {
                next[static_cast<std::size_t>(syy) * w + sxx] =
                    grid_[static_cast<std::size_t>(dyy) * w + dxx];
            }
        }
        grid_ = std::move(next);
    } else {
        clear();
    }
    ox = new_ox;
    oy = new_oy;
}

std::pair<int, int> OccupancyGrid::cell(double x, double y) const {
    // static_cast<int> truncates toward zero — same as Python int().
    return {static_cast<int>((x - ox) / res),
            static_cast<int>((y - oy) / res)};
}

bool OccupancyGrid::in_bounds(double x, double y) const {
    const auto [cx, cy] = cell(x, y);
    return cx >= 0 && cx < w && cy >= 0 && cy < h;
}

// ------------------------------------------------------------------ update

void OccupancyGrid::bump(int cx, int cy, double delta) {
    if (cx >= 0 && cx < w && cy >= 0 && cy < h) {
        float& v = grid_[static_cast<std::size_t>(cy) * w + cx];
        v = static_cast<float>(
            clampf(v + delta, lo_min, lo_max));
    }
}

void OccupancyGrid::mark(double x, double y, bool occupied) {
    const auto [cx, cy] = cell(x, y);
    bump(cx, cy, occupied ? lo_hit : lo_free);
}

void OccupancyGrid::mark_points(const std::vector<Vec3>& points, bool occupied) {
    for (const Vec3& p : points) mark(p.x, p.y, occupied);
}

void OccupancyGrid::raycast(double rox, double roy, double x, double y,
                            bool hit) {
    auto [cx0, cy0] = cell(rox, roy);
    auto [cx1, cy1] = cell(x, y);
    const int dx = std::abs(cx1 - cx0);
    const int dy = -std::abs(cy1 - cy0);
    const int sx = cx0 < cx1 ? 1 : -1;
    const int sy = cy0 < cy1 ? 1 : -1;
    int err = dx + dy;
    int cx = cx0, cy = cy0;
    while (true) {
        if (cx == cx1 && cy == cy1) {
            bump(cx, cy, hit ? lo_hit : lo_free);
            break;
        }
        bump(cx, cy, lo_free);
        const int e2 = 2 * err;
        if (e2 >= dy) {
            err += dy;
            cx += sx;
        }
        if (e2 <= dx) {
            err += dx;
            cy += sy;
        }
    }
}

void OccupancyGrid::insert_scan(const std::vector<Vec3>& points,
                                std::pair<double, double> origin,
                                std::optional<double> max_range) {
    const double sox = origin.first, soy = origin.second;
    for (const Vec3& p : points) {
        if (max_range &&
            std::hypot(p.x - sox, p.y - soy) > *max_range)
            continue;
        raycast(sox, soy, p.x, p.y, true);
    }
}

// ------------------------------------------------------------------- query

double OccupancyGrid::log_odds(double x, double y) const {
    const auto [cx, cy] = cell(x, y);
    if (cx >= 0 && cx < w && cy >= 0 && cy < h)
        return static_cast<double>(grid_[static_cast<std::size_t>(cy) * w + cx]);
    return 0.0;
}

bool OccupancyGrid::is_occupied(double x, double y, double threshold) const {
    const auto [cx, cy] = cell(x, y);
    return cx >= 0 && cx < w && cy >= 0 && cy < h &&
           grid_[static_cast<std::size_t>(cy) * w + cx] > threshold;
}

double OccupancyGrid::occupied_fraction() const {
    if (grid_.empty()) return 0.0;
    std::size_t count = 0;
    for (float v : grid_)
        if (v > 0.6f) ++count;
    return static_cast<double>(count) / static_cast<double>(grid_.size());
}

double OccupancyGrid::free_space_ahead(double ego_x, double ego_y, double yaw,
                                       double max_dist,
                                       double half_width) const {
    const double ux = std::cos(yaw), uy = std::sin(yaw);
    const double vx = -uy, vy = ux;  // lateral unit vector
    const int n_lat = std::max(1, static_cast<int>(std::ceil(half_width / res)));
    // lats = linspace(-half_width, half_width, 2*n_lat + 1)
    std::vector<double> lats(static_cast<std::size_t>(2 * n_lat + 1));
    for (int i = 0; i <= 2 * n_lat; ++i)
        lats[static_cast<std::size_t>(i)] =
            -half_width + (2.0 * half_width) * i / (2.0 * n_lat);
    const double step = res;
    double s = step;
    while (s < max_dist) {
        const double bx = ego_x + ux * s, by = ego_y + uy * s;
        if (!in_bounds(bx, by)) return s;  // ran out of known space
        for (double lat : lats) {
            if (is_occupied(bx + vx * lat, by + vy * lat)) return s;
        }
        s += step;
    }
    return max_dist;
}

std::vector<std::uint8_t> OccupancyGrid::as_uint8() const {
    std::vector<std::uint8_t> out(grid_.size());
    for (std::size_t i = 0; i < grid_.size(); ++i)
        out[i] = static_cast<std::uint8_t>(
            clampf(128.0 + static_cast<double>(grid_[i]) * 64.0, 0.0, 255.0));
    return out;
}

}  // namespace fsd
